src/main/java/com/exactpro/th2/codec/qfj/QFJCodec.java

```diff
@@ -0,0 +1,473 @@
+        fieldsMap.put("Header", ValueUtils.toValue(header));
+
+        Iterator<Field<?>> iterator = qfjMessage.iterator();
+        fillMessageBody(iterator, fieldsMap, qfjMessage, msgType);
+
+        Iterator<Field<?>> trailerIterator = qfjMessage.getTrailer().iterator();
+        Message trailer = getMessage(trailerIterator, Message.newBuilder(), strMessage);
+        fieldsMap.put("Trailer", ValueUtils.toValue(trailer));
+
+
+        String msgName = localDataDictionary.getValueName(MsgType.FIELD, msgType);
+        return Message.newBuilder()
+                .putAllFields(fieldsMap)
+                .setParentEventId(rawMessage.getParentEventId())
+                .setMetadata(MessageMetadata.newBuilder()
+                        .setId(rawMessage.getMetadata().getId())
+                        .setTimestamp(rawMessage.getMetadata().getTimestamp())
+                        .setProtocol(rawMessage.getMetadata().getProtocol())
+                        .setMessageType(msgName)
+                        .putAllProperties(rawMessage.getMetadata().getPropertiesMap())
+                        .build())
+                .build();
+    }
+
+    private void fillMessageBody(Iterator<Field<?>> iterator, Map<String, Value> fieldsMap, quickfix.Message qfjMessage, String msgType) {
+        while (iterator.hasNext()) {
+            Field<?> field = iterator.next();
+
+            if (localDataDictionary.isGroup(msgType, field.getTag())) {
+                List<Group> groups = qfjMessage.getGroups(field.getTag());
+                ListValue.@NotNull Builder listValue = ValueUtils.listValue();
+                DataDictionary innerDataDictionary = localDataDictionary.getGroup(msgType, field.getTag()).getDataDictionary();
+
+                fillListValue(listValue, innerDataDictionary, groups, msgType);
+                fieldsMap.put(localDataDictionary.getFieldName(field.getTag()), ValueUtils.toValue(listValue));
+            } else {
+                if (!localDataDictionary.isMsgField(msgType, field.getField())) {
+                    throw new IllegalArgumentException("Invalid tag \"" + localDataDictionary.getFieldName(field.getField()) + "\" for message type " + msgType);
+                }
+                putField(localDataDictionary, fieldsMap, field);
+            }
+        }
+
+    }
+
+    private void fillListValue(ListValue.Builder listValue, DataDictionary dataDictionary, List<Group> groups, String msgType) {
+        for (Group group : groups) {
+            Message.Builder messageBuilder = Message.newBuilder();
+            Iterator<Field<?>> innerIterator = group.iterator();
+            Message innerMessage = getMessage(innerIterator, messageBuilder, dataDictionary, group, msgType);
+            listValue.addValues(ValueUtils.toValue(innerMessage));
+        }
+    }
+
+    private Message getMessage(Iterator<Field<?>> iterator, Message.Builder messageBuilder, DataDictionary dataDictionary, Group group, String msgType) {
```

Comment:
> Message.Builder messageBuilder

Why do we need it? In all calls we just pass a newly created builder which isn't used anywhere outside this method